### PR TITLE
Enriching current statistics framework

### PR DESCRIPTION
**Pull request by ccjeremylo**
### Change:
- Record different types of nodes visited
- Record number of pruning's that took place
- Record number of times when TT was used to return node score
- Added trivial unit tests for statistics behaviour

Ideally would like to also:
- record different types of pruning's, and
- add more tests, specifically checking if the searcher is indeed incrementing the statistics accurately

can decide later if we want to do in this PR.

### Context: 
Previously, the statistics class only records number of nodes visited, and did not distinguish between the nature of the nodes. For performance and (more generally) behavioural analysis, we would like to be able to record:
- different types of nodes visited
- number of prunings that took place
- number of times when TT was used to return node score

See #99 for more background

### CODE REVIEW

File: sporkfish/searcher/searcher.py
Comment (KYLChiu) [inline sporkfish/searcher/searcher.py:60]: For future, perhaps this deserves to be part of the Stats API?

**Review by KYLChiu — APPROVED** [review]
Nice one - just a couple of discussion points imo but won't block. Thanks for cleaning up tests also!

File: sporkfish/searcher/searcher.py
Comment (ccjeremylo) [reply sporkfish/searcher/searcher.py:60]: by "this" do you mean the `NodeTypes` enum?

File: sporkfish/statistics.py
Comment (ccjeremylo) [reply sporkfish/statistics.py:28]: Reason I did it this way was because I felt TT & pruning are logically very distinct from the node counts. The distinction makes it clear to readers. 

I want to enrich the pruning count to also be a dictionary in the future, categorising different types of pruning. For this reason, I think they deserve different interfaces?

Agree with you the interface can be designed better in general.

File: tests/test_statistics.py
Comment (ccjeremylo) [inline tests/test_statistics.py:4]: This was part of the original ask in #99. Maybe I do it in next PR - I do think it is very important to have tests to check that searcher is incrementing the stats correctly.

**Review by ccjeremylo — COMMENTED** [review]
Was not expecting such a quick approve @KYLChiu 😸 But still can't merge yet, think I must have pushed something after your approval.

Care to reapprove?

File: sporkfish/searcher/searcher.py
Comment (KYLChiu) [reply sporkfish/searcher/searcher.py:60]: I think the logic to print a the stats can live in the API, maybe even as __str__.

File: sporkfish/statistics.py
Comment (KYLChiu) [reply sporkfish/statistics.py:28]: Im not sure readability is a big concern because when they call the API they need to pass in the enum to increment anyway, like increment(NodeType.QUIESCENCE) or increment(Pruning). I think it mainly becomes an issue when we try to aggregate these, the logic becomes slightly more complicated. Anyway no strong opinions either way

**Pull request merged**
